### ISSUE TITLE
Multipart with JSON

### ISSUE THREAD

**Issue by magalhas** [visionmedia/superagent#455]
Hi,

Can you please let me know how to attach a file and send the rest of the data as JSON? I've tried using req.part() but it has been deprecated.

Thanks.

**Comment by zbyte64** [comment]
Okay two things,

Is `part` really deprecated? If so shouldn't it be removed from the docs?

Can attach accept a file stream or a blob of text isntead of reading from a filename?

**Comment by formula1** [comment]
It appears this was not answered

Heres what I have found if it helps anyone

* [superagent](https://github.com/visionmedia/superagent/blob/master/lib/node/index.js#L192) basically is just a wrapper around the [form-data](https://github.com/felixge/node-form-data) module when it comes to multipart post calls. The wrapper transforms strings into readable streams so your best to expect that a normal stream should work as well
* [form-data](https://github.com/felixge/node-form-data/blob/master/lib/form_data.js#L23)'s append method is a wrapper around [combined-stream](https://github.com/felixge/node-combined-stream)'s append. 
* [combined-stream](https://github.com/felixge/node-combined-stream/blob/master/lib/combined_stream.js#L38)'s converts all streams into a delayed stream then pushes the value (regardless of whether its a stream or not) on the list of streams  
* [combined-stream](https://github.com/felixge/node-combined-stream/blob/master/lib/combined_stream.js#L95)'s way of handling its list of streams is pretty straight forward. If its a stream, it pipes. If it isn't it just writes the values.

In other words passing a blob or a readableStream seems to be fine.

If you want to pass a content type, you must provide a exstension that will qualify as the type you want. This seems straightforward however it is not in all cases since if you want to send a file as `application/x-patch` that will not be possible as it is not in the [`mime-db`](https://github.com/jshttp/mime-db/blob/master/db.json)

**Comment by misha-slyusarev** [comment]
I'm using superagent with React and Rails and trying to compose a multipart request for the backend.

Is there a way to add a JS object to the request without adding it field by field? For instance, I would like to have `{'invoice' => {'recipient_attributes' => {'name' => 'myname', 'last_name' => 'mylastname'}, 'attachment' => ... }` as my params in Rails backend part. And that's how I'm trying to do it:
```
request.post('v1/invoices')
      .field('invoice[recipient_attributes]', this.state.recipient)
      .attach('invoice[attachment]', this.state.invoiceFile)
      .end(function(err, res){
        console.log('invoiceInfo submitted')
      });
```

But `recipient_attributes` is saved as '[object Object]' instead of expected. Is there a way to add the whole object, or I have to add fields one by one like this:
```
.field('invoice[recipient_attributes][name]', 'myname')
.field('invoice[recipient_attributes][last_name]', 'mylastname')
```

Thank you in advance!

**Comment by pornel** [comment]
You have to add fields one by one.

**Comment by ollydixon** [comment]
The fields get converted to strings and fail validation.
